Implementation of IssnSubjectCodes for Crossref-2024 with download method

panos-span opened the pull request:
**Title:** Implementation of IssnSubjectCodes for Crossref-2024, introducing a new method for `data_source` called `download`.

**Description:**
This pull request introduces the `IssnSubjectCodes` module for Crossref-2024 into the Alexandria3k project. The module processes ISSN to ASJC (All Science Journal Classification) subject codes, supporting queries over its virtual table and enabling the population of an SQLite database with its data. This is a great work-around in order to get the subjects for the works table since crossref doesn't support them anymore.

Key features of this implementation include:
- Definition of the `IssnSubjectCodes` class which creates an object to handle ISSN to ASJC meta-data.
- A method to populate an SQLite database with ISSN and subject codes data, with an initial step to create the required CSV from API calls.
- Mock tests for validating the CSV creation and database population processes.
- Created the `download` command for `data_source.py` and `__main__.py`

**Changes:**
- Added `issn_subject_codes.py` to `src/alexandria3k/data_sources/` with the implementation of the `IssnSubjectCodes` class.
- Added SQL script `get-issns.sql` to `src/alexandria3k/sql/` for retrieving ISSNs from the `works` table.
- Added a test CSV `issn_subjects.csv` to `tests/data/` for unit tests.
- Added unit tests in `test_issn_subject_codes.py` under `tests/data_sources/` to validate the module's functionality.

**Commit Details:**
- **src/alexandria3k/data_sources/issn_subject_codes.py:**
  - Introduced `IssnSubjectCodes` class.
  - Added method to create CSV from API calls and create a data folder where the script is executed.
  - Added populate method override to handle CSV creation and data insertion into the database.

- **src/alexandria3k/sql/get-issns.sql:**
  - Added SQL script to select distinct ISSNs from the `works` table.

- **tests/data/issn_subjects.csv:**
  - Added sample CSV data for unit testing.

- **tests/data_sources/test_issn_subject_codes.py:**
  - Added unit tests to validate the `IssnSubjectCodes` module's CSV creation and database population.

**Library Used:**
This implementation utilizes the `pybliometrics` library to interact with the Elsevier API. The ASJC codes are retrieved through this API. Users will need an API key and an institutional token to access the Elsevier API. The config file `pybliometrics.cfg` must be inside the `./config` folder of the user.

**Testing:**
Unit tests have been added to verify:
- CSV creation with correct ISSN and subject code entries.
- Correct population of the SQLite database with the generated CSV data.

**Notes:**
Ensure that the `pybliometrics` library is correctly initialized and that the required API keys are configured.

**Checklist:**
- [x] Code adheres to the project's style guidelines.
- [x] Documentation and comments are included as needed.
- [x] Unit tests are added and passing.

dspinellis commented:
This is impressive; well done!  Please:

* [ ] Fix the CI failure
* [ ] Add some indication of the time required to download the data
* [ ] Consider splitting the process into the formation of the data set (with a new action `download` verb and directory for its plugins) and the traditional population step.

panos-span commented:
The API lets you make 9 requests per second, with a limit of 25000 requests per week, when having an institutional token.

This data source should only be applied for custom examples, when a user has already in mind the range of the issn subjects that might be needed.

dspinellis commented on src/alexandria3k/data_sources/issn_subject_codes.py:
```diff
@@ -102,11 +107,50 @@ def __init__(
             attach_databases,
         )
+        
+    def download_data(self, database_path):
```
I guess the database path isn't needed, but a file name might, right?
You also need to provide a CLI way for calling this.

panos-span replied:
I believe the database path is needed since the user must define it in order to get the ISSNs from the wanted works table at the database he needs.

dspinellis replied:
Ah, good point!  I now see there's an implicit dependency on the `works` table and some fields, which you should document.  Ideally, the SQL query should be specified by the user.  Please allow the specification of the file where the data are written.  Alternatively, you could populate the table on the fly, without an intermediate CSV file.

dspinellis commented on src/alexandria3k/data_sources/issn_subject_codes.py:
```diff
@@ -0,0 +1,156 @@
+    get_string_resource,
+)
+
+DEFAULT_SOURCE = "resource:data/issn_subject_codes.csv"
```
Please make this an API/CLI argument.  See e.g. Crossref.

dspinellis reviewed (changes requested):
Getting there!  Please have a look at the comments below. Mark the check-boxes of what you have finished, and ping me for another review. Also, consider squashing your commits as needed.

panos-span commented:
I implemented the recommended changes. I just don't know where to upload a test config file for pybliometrics so the CI will pass. 

Once this will be addresed, i will squash my commits.

dspinellis commented:
> I implemented the latest recommended changes. I just don't know where to upload a test config file for pybliometrics so the CI will pass.
> 
> Once this will be addresed, i will squash my commits.

The CI script can copy the config file from `tests/data` to  `$HOME`. However, note that the config file shall not include any secret (e.g. API) keys.  If secrets are needed to test it, then you should mock the response.

panos-span commented:
alexandria3k/data_sources/issn_subject_codes.py:89:4: R0913: Too many arguments (6/5) (too-many-arguments)
 
 Should i give less options to the user? What should i do?

dspinellis commented:
> alexandria3k/data_sources/issn_subject_codes.py:89:4: R0913: Too many arguments (6/5) (too-many-arguments)
> 
> Should i give less options to the user? What should i do?
> 

* Add a new subcommand `download` which looks in the new `download_sources` for plugins.  It can have as a positional argument the output data path and as an optional arguments an SQL query to yield the keys to download.
* Use `tempfile.NamedTemporaryFile` to obtain a temporary file path where the data will be stored. Delete this when the database gets populated.

panos-span commented:
Everything should work as intended now!

dspinellis commented on src/alexandria3k/__main__.py:
```diff
@@ -1,536 +1,569 @@
+        "download", help="Download data using the specified data source."
+    )
+    parser.set_defaults(func=download)
+    parser.add_argument(
```
This should be optional.

dspinellis commented on src/alexandria3k/__main__.py:
```diff
@@ -1,536 +1,569 @@
+        choices=facility_names("data_sources"),
+        help="Name of the data source to use",
+    )
+    parser.add_argument(
```
This should be optional (an option) where both the database and the SQL query should be given or not given.

dspinellis commented on src/alexandria3k/__main__.py:
```diff
@@ -1,536 +1,569 @@
+    parser = subparsers.add_parser(
+        "download", help="Download data using the specified data source."
+    )
+    parser.set_defaults(func=download)
```
I think the should be an unspecified number of positional arguments depending on the source, e.g. URL, key, file to retrieve.

panos-span replied:
```python
parser.add_argument(
    "extra_args",
    nargs="*",
    help="Additional arguments for the data source (e.g. URL, key, file path)",
)
```

Will this suffice?

dspinellis replied:
Should do.  Each data source will verify that the correct arguments are passed.

dspinellis commented on src/alexandria3k/__main__.py:
```diff
@@ -1,536 +1,569 @@
+    parser = subparsers.add_parser(
+        "download", help="Download data using the specified data source."
+    )
+    parser.set_defaults(func=download)
```
Consider adding an option for the output directory or file.

panos-span replied:
The output file in our case, is temporary. Should we proceed with that?

dspinellis replied:
Since you added a `download` subcommand, wouldn't the workflow be as follows?
1. download
2. populate

panos-span replied:
Yes correct.

dspinellis commented on src/Pipfile:
```diff
@@ -9,6 +9,9 @@ importlib-metadata = "*"
 pyahocorasick = "*"
 
 [dev-packages]
+tqdm = "*"
```
We already have a progress meter. Please use that one rather than introducing another dependency

dspinellis commented on src/Pipfile:
```diff
@@ -9,6 +9,9 @@ importlib-metadata = "*"
 pyahocorasick = "*"
 
 [dev-packages]
```
Why are these "Dev" packages? I think they are required at runtime.

dspinellis commented on src/Pipfile:
```diff
@@ -9,6 +9,9 @@ importlib-metadata = "*"
 
 [dev-packages]
+tqdm = "*"
+pandas = "*"
```
Pandas is a very big and heavy dependency. Do we actually need it? Can we avoid it?

panos-span replied:
Yes i believe we can.

dspinellis commented on src/alexandria3k/__main__.py:
```diff
@@ -1,536 +1,569 @@
-#!/usr/bin/env python3
```
The whole file has had its whitespace changed (I guess line endings) and appears completely changed under `git diff`.  Please adjust Git settings to and the line endings to avoid this.

dspinellis commented on src/alexandria3k/data_sources/issn_subject_codes.py:
```diff
@@ -0,0 +1,197 @@
+#
+# Alexandria3k Crossref bibliographic metadata processing
+# Copyright (C) 2022-2023  Diomidis Spinellis
```
Add you name 2024 here.

dspinellis commented on src/alexandria3k/data_sources/issn_subject_codes.py:
```diff
@@ -0,0 +1,197 @@
+                raise FileNotFoundError(
+                    f"Configuration file not found at {config_path}"
+                )
+            print(f"Using config file at {config_path}")
```
Make this debug output. The program should be silent by default.

dspinellis commented on src/alexandria3k/data_sources/issn_subject_codes.py:
```diff
@@ -0,0 +1,197 @@
+
+        # Set the configuration file path
+        config_path = get_config_path(config_path)
+        os.environ["PYBLIOMETRICS_CONFIG_FILE"] = config_path
```
Shouldn't you use this environment variable if set?

dspinellis commented on src/alexandria3k/data_sources/issn_subject_codes.py:
```diff
@@ -0,0 +1,197 @@
+
+    def fetch_subject_codes(self, writer, issns):
+        """Fetch the subject codes for the specified ISSNs."""
+        for issn in tqdm(issns, desc="Processing ISSNs"):
```
`Downloading`, not `Processing`, right?

dspinellis reviewed (changes requested):
Overall it's going in the right direction; well done!  Some more work is needed, mostly in the area of the download API definition.  It should be flexible enough to cover most possible uses; consider e.g. Crossref or US patents.

dspinellis commented:
Note that a data source may (as in your case) or may not (as in all others currently) support a download subcommand.  In the second case a suitable error message should be shown.

panos-span commented:
- [x] Added requested options for user for `download` command
- [x] Added necessary handling of workflow operation of `download` with `populate`
- [x] Removed unnecessary libraries
- [x] Created progress bar for ISSN downloading
- [x] Fixed white space `git diff` issue

dspinellis commented on src/alexandria3k/__main__.py:
```diff
@@ -99,10 +99,82 @@ def get_data_source_instance(args):
+    if hasattr(data_source_instance, "download"):
+        # Check if the ouput_path attribute is not None
+        if data_source_instance.output_path is None:
+            raise Alexandria3kError(
```
Please also take into account the CLI use. There the program is invoked twice: to download and then to populate.

dspinellis replied:
I am referring to the [CLI](https://dspinellis.github.io/alexandria3k/cli-eg.html), not the [API](https://dspinellis.github.io/alexandria3k/python-eg.html).

panos-span replied:
Hmmm, you got a point, should we do separate checks for the API (like the one we have) and the CLI , or should we do it with an another way? What do you suggest?

panos-span replied:
I am a little confused, should we check if both are called in the same time? For the `download` we give the database path to create the tables into and for the `populate` function to just populate in the same database. At least in our case, it isn't mandatory to first call the `download` function, it just makes more sense.

dspinellis reviewed (changes requested):
Excellent handling of the previous round comments; well done!  The CLI API needs some more thought.

panos-span commented:
Now, we can accept for the `populate` method as input the `input_path` , as well as for the `download` method.
In addition to that, we also check if the populate method signature can take the extra parameter so there wont be any occurring issues. Lastly, the file created is no longer temporary.

dspinellis commented on src/alexandria3k/data_sources/issn_subject_codes.py:
```diff
@@ -0,0 +1,218 @@
+            return config_path
+
+        # Check if the environment variable is set or not
+        if "PYBLIOMETRICS_CONFIG_FILE" not in os.environ:
```
This sequence should be before the previous `if`, no?

dspinellis commented on src/alexandria3k/data_sources/issn_subject_codes.py:
```diff
@@ -0,0 +1,218 @@
+    its (virtual) table and the population of an SQLite database with its
+    data.
+
+    :param data_source: The location (file path or URL) where the data
```
Only file path, not URL, right?
Also, "data are to be stored (for download) or loaded from (for populate or query)"

panos-span replied:
Yes correct.

dspinellis commented on src/alexandria3k/__main__.py:
```diff
@@ -99,19 +100,102 @@ def get_data_source_instance(args):
+        type=str,
+        help="SQL query to retrieve the data for downloading",
+    )
+    parser.add_argument(
```
This should be a required positional argument to be consistent with the `populate` method.

dspinellis commented on src/alexandria3k/data_sources/issn_subject_codes.py:
```diff
@@ -0,0 +1,218 @@
+        """
+        Create a CSV file with ISSNs and their corresponding ASJC subject codes from API calls.
+        """
+        if output_path is None:
```
This would be the `data_source` field, no?

dspinellis commented on src/alexandria3k/data_sources/issn_subject_codes.py:
```diff
@@ -0,0 +1,218 @@
+        population, defaults to `None`.
+        :type condition: str, optional
+
+        :param input_file_path: The path specifying the CSV file to use for the population,
```
No need for this, it is the `data_source` field.

panos-span replied:
If i remember correctly, you mentioned that we want to pass as input the data_source, am i right?

dspinellis reviewed (changes requested):
We're closing in! I think the code will be much more compact and clean after these changes.

Also, please add in `data_source.py` a default `download` method, which should through an unsupported operation exception.

panos-span commented:
- [x] Added in `data_source.py` a default `download` method
- [x] Adjusted `data_source` usage

dspinellis commented on src/alexandria3k/data_sources/issn_subject_codes.py:
```diff
@@ -0,0 +1,221 @@
+                    writer.writerow([issn, code])
+
+            except (KeyError, ValueError) as e:
+                print(f"Error processing ISSN {issn}: {e}")
```
`raise Alexandria3kError(f"…") from e`

panos-span replied:
I believe we should be able to bypass this without the program stopping.

dspinellis replied:
Good point! Then output this as a warning on the standard error channel by calling `common.warn()`

dspinellis commented on src/alexandria3k/data_sources/issn_subject_codes.py:
```diff
@@ -0,0 +1,221 @@
+        try:
+            pybliometrics.scopus.init()
+        except RuntimeError as e:
+            raise e
```
raise Alexandria3kError("Error in downloading data with pybliometrics") from e

dspinellis commented on src/alexandria3k/data_sources/issn_subject_codes.py:
```diff
@@ -0,0 +1,221 @@
+                        subject_area_codes = result[
+                            "subject_area_codes"
+                        ].split(";")
+                        subject_area_codes_set.update(subject_area_codes)
```
Do you have to store everything in memory and then write it? Can't you stream it all directly to the file?

dspinellis commented on src/alexandria3k/data_sources/issn_subject_codes.py:
```diff
@@ -0,0 +1,221 @@
+        :param sql_query: The SQL query to use in order to fetch the ISSNs, defaults to `None`.
+        :type sql_query: str, optional
+        """
+        data_source = os.path.join(data_source, "issn_subject_codes.csv")
```
No need to append a fixed file name. Let the user specify it.

dspinellis commented on src/alexandria3k/data_sources/issn_subject_codes.py:
```diff
@@ -0,0 +1,221 @@
+        :param data_source: The path specifying the CSV file to use for the population.
+        :type data_source: str
+
+        :param sql_query: The SQL query to use in order to fetch the ISSNs, defaults to `None`.
```
Document what the query needs and what it shall return, and what the default behavior is.

dspinellis reviewed (changes requested):
Still work to do I'm afraid. I hope this doesn't discourage you.

panos-span commented:
> Still work to do I'm afraid. I hope this doesn't discourage you.

No problem. Every correction is welcome and appreciated.

dspinellis commented on src/alexandria3k/data_sources/issn_subject_codes.py:
```diff
@@ -0,0 +1,198 @@
+        :param data_location: The path specifying the CSV file to use for the population.
+        :type data_location: str
+
+        :param sql_query: The SQL query to use in order to fetch the ISSNs, defaults to `None`.
```
"The SQL query that will supply the ISSNs to fetch"

dspinellis commented on src/alexandria3k/data_sources/issn_subject_codes.py:
```diff
@@ -0,0 +1,198 @@
+        :type data_location: str
+
+        :param sql_query: The SQL query to use in order to fetch the ISSNs, defaults to `None`.
+            The default query fetches all unique ISSNs from the database. The query should
```
Clarify, from which database and which column(s).

dspinellis commented on src/alexandria3k/data_sources/issn_subject_codes.py:
```diff
@@ -0,0 +1,198 @@
+        new_data_source = VTSource(
+            issn_subject_codes_table, data_location, self.sample
+        )
+        self.data_source = new_data_source
```
I think you can handle the above lines in `__init__` and dispense with the `populate` method entirely.

panos-span replied:
I agree. However, i left it for the `data_location` parameter, in case the user wants to modify the parameter from `__init__`.

dspinellis replied:
The CLI should also be able to work with the constructor API, but passing it the appropriate arguments for each subcommand, no?

dspinellis replied:
I had a typo in my comment: "but passing" → "by passing".

Both populate and download should work.

Please try a small example through the CLI to make sure.

panos-span replied:
I now removed the populate method from `IssnSubjectCodes`. I am testing now some cases from the CLI.

panos-span replied:
You were correct (very unexpected, i know haha) ! I tested it and it works just as you said. I also added some necessary arguments that i had forgot.

dspinellis reviewed (changes requested):
Now the structure of your contribution and the API are very clear, so most requested changes are cosmetic.  Well done; almost there!

panos-span commented on src/alexandria3k/data_sources/issn_subject_codes.py:
```diff
@@ -0,0 +1,198 @@
+import os
+import sqlite3
+import pybliometrics
+from pybliometrics.scopus import SerialSearch
```
src/alexandria3k/data_sources/issn_subject_codes.py:29:0: C0411: standard import "sqlite3" should be placed before third party imports "pybliometrics", "pybliometrics.scopus.SerialSearch" (wrong-import-order)

dspinellis replied:
Ah, OK.  Please separate local from package from standard imports with a blank line.

dspinellis commented on src/alexandria3k/data_sources/issn_subject_codes.py:
```diff
@@ -131,19 +132,24 @@ def fetch_subject_codes(self, writer, issns):
-            return a single column with the name `issn`.
+        :param sql_query: The SQL query that will supply the ISSNs to fetch,
+        from the database specified, where the ISSNs are represented by `issn_print`
+        and `issn_electronic`, defaults to `None`.
```
Please remove "where the ISSNs are represented by `issn_print` and `issn_electronic`" as this is not a required part of the specified interface. Please also clarify that the ISSN shall be formatted as 8 alphanumeric characters (.e.g `2249782X`) without any spaces or dashes.

dspinellis commented on src/alexandria3k/__main__.py:
```diff
@@ -99,10 +99,83 @@ def get_data_source_instance(args):
+    parser.add_argument(
+        "data_location",
+        type=str,
+        help="File path to save the downloaded data",
```
"File or directory path" The crossref download will download to a directory.

panos-span commented:
All done!

dspinellis reviewed (approved):
Well done! We can merge now. Thank you for your contribution.
